en/lessons/basics/containers.md

```diff
@@ -0,0 +1,799 @@
+ghci> List.lookup "foo" assoc
+Just True
+ghci> List.lookup "bar" assoc
+Just False
+ghci> List.lookup "baz" assoc
+Nothing
+```
+
+We can see the pattern, again, where no result is encoded using the `Maybe`
+type. This is because it is always *possible* that the key are looking up isn't
+present in our assoc list.
+
+It is also interesting to note the `Eq a` constraint on the "key" which allows
+the lookup function to do an equality comparison on them.
+
+While assoc lists are a nice introduction to key value containers, and they build
+on the previous types we learned about, they are not particularly useful. A list
+simply isn't a very good data structure for lookup, as it provides worst case
+`𝛰(n)` asymptotics. Assoc lists are usually an intermediate data structure
+which Haskell programmers will convert into a `Map`. Although this
+conversion is itself an `𝛰(n*log n)` operation `Map` provides an `O(log n)`
+lookup, so the conversion cost is amortised very quickly.
+
+## Sets
+
+Sets are a very interesting container, the core concept of a set is membership.
+It is common to build up sets, and then test to see if an element exists in that
+set.
+
+A set can be constructed exclusively by inserting elements into the empty set
+
+```haskell
+ghci> :t Set.empty
+Set.empty :: Set.Set a
+ghci> Set.empty
+fromList []
+ghci> Set.insert 1 (Set.insert 2 (Set.insert 3 Set.empty))
+fromList [1,2,3]
+```
+
+Or by creating a set from a list
+
+```haskell
+ghci> Set.fromList [4,3,2,1]
+fromList [1,2,3,4]
+```
+
+You might notice that the elements are sorted after turning them into a `Set`.
+Internally the `Set` data type depends on its contents being orderable. The
+concrete implementation of sets in Haskell are binary trees, which depend on an
+ordering. We can see the ordering requirement from the constraints on the
+`insert` and `fromList` (and other) functions.
+
+```haskell
+ghci> :t Set.insert
+Set.insert :: Ord a => a -> Set.Set a -> Set.Set a
+ghci> :t Set.fromList
+Set.fromList :: Ord a => [a] -> Set.Set a
+ghci> :t Set.member
+Set.member :: Ord a => a -> Set.Set a -> Bool
+```
+
+Sets have a very useful property, they cannot contain duplicates.
+
+```haskell
+ghci> insert1 = Set.insert 1
+ghci> insert1 Set.empty
+fromList [1]
+ghci> insert1 (insert1 Set.empty)
+fromList [1]
+ghci> insert1 (insert1 (insert1 Set.empty))
+fromList [1]
+```
+
+Alright, let's see an almost-practical use case for sets.
+
+```haskell
+ghci> evens = Set.fromList [0,2..1000000]
+ghci> Set.member 7 evens
+False
+ghci> Set.member 200012 evens
+True
+ghci> isEven n = Set.member n evens
+ghci> isEven 7
+False
+ghci> isEven 8
+True
+```
+
+You might say "hmmm a `1000000` limit for even numbers seems incorrect", and you
+would be right! This highlights a property of sets in haskell, they are finite
+due to the strictness in the underlying data structure. This contrasts lists,
+which are lazy and therefore potentially infinite.
+
+### Difference
```

Comment:
This series of set-related operations take incredibly large space for such a niche instrument: I rarely find myself taking set differences, for example. I think 20% of interface (really: insert and lookup) cover 80% of users.

More concretely, I’d suggest to pack the set-theoretic operations (diff, union, intersect and subset; maybe Cartesian product) into one section with a single block of examples. This is a drastic change, so maybe if others are really happy with the current version, I don’t want to insist. But I wholeheartedly believe this topic is currently vastly overweighted with little benefit.

Reply:
Would be interested to hear what others think

Reply:
Shortening the page (by grouping operations together and using more terse language) would likely make it more digestible for readers and increase the speed of the progress they make.

Reply:
As said in my other review, I would be absolutely delighted if the Set part could be a fully-fledged blog post on Haskell School. The point of these lessons is to give a touch and feel of the language, sometimes at the price of exhaustiveness.

Reply:
It seems there is consensus!